src/compiler/parser.cc

```diff
@@ -1048,14 +1054,36 @@ Class* Parser::parse_class_interface_monitor_or_mixin(bool is_abstract) {
   } else {
     name = parse_identifier();
     bool requires_super = false;
-    if (current_token() == Token::IDENTIFIER && current_token_data() == Symbols::extends) {
+    if (at(Symbols::extends)) {
       consume();
       requires_super = true;
     }
-    if (current_token() == Token::IDENTIFIER && current_token_data() != Symbols::implements) {
+
+    if (current_token() == Token::IDENTIFIER && !at(Symbols::implements) && !at(Symbols::with)) {
       super = parse_type(false);
     }
-    if (current_token() == Token::IDENTIFIER && current_token_data() == Symbols::implements) {
+    if (at(Symbols::with)) {
+      if (!requires_super) {
+        // "requires_super" indicates that there was an `extends`.
+        // TODO(florian): do we really want this requirement.
+        // The `class A with Foo` reads a bit worth than `class A extends Object with Foo`, but
```

Comment:
worth? I don't understand the comment as is.

Reply:
Removed the TODO and comment. We can always change it, if we feel like it.
Comment was saying that `class A with Foo` isn't really English (compared to `class A extends B with Foo`), but that we would probably still be ok with it.
done.